model/log/sender.go

```diff
@@ -0,0 +1,210 @@
+// the backing log service. Send is thread safe.
+func (s *sender) Send(m message.Composer) {
+	ts := time.Now().UnixNano()
+
+	if !s.Level().ShouldLog(m) {
+		return
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.closed {
+		s.opts.Local.Send(message.NewErrorMessage(level.Error, errors.New("cannot call Send on a closed sender")))
+		return
+	}
+
+	for _, line := range strings.Split(m.String(), "\n") {
+		if line == "" {
+			continue
+		}
+
+		logLine, err := s.opts.Parse(line)
+		if err != nil {
+			s.opts.Local.Send(message.NewErrorMessage(level.Error, errors.Wrap(err, "parsing log line")))
+			return
+		}
+		if logLine.Priority == 0 {
+			logLine.Priority = m.Priority()
+		}
+		if !logLine.Priority.IsValid() {
+			s.opts.Local.Send(message.NewErrorMessage(level.Error, errors.Errorf("invalid log line priority %d", logLine.Priority)))
+			return
+		}
+		if logLine.Timestamp == 0 {
+			logLine.Timestamp = ts
+		}
+		if logLine.Timestamp < 0 {
+			s.opts.Local.Send(message.NewErrorMessage(level.Error, errors.Errorf("invalid log line timestamp %d", logLine.Timestamp)))
+			return
+		}
+
+		s.buffer = append(s.buffer, logLine)
+		s.bufferSize += len(line)
+		if s.bufferSize > s.opts.MaxBufferSize {
+			if err := s.flush(s.ctx); err != nil {
+				s.opts.Local.Send(message.NewErrorMessage(level.Error, err))
+				return
+			}
+		}
+	}
+}
+
+// Flush flushes anything messages that may be in the buffer to persistent
+// storage determined by the backing log service.
+func (s *sender) Flush(ctx context.Context) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.closed {
+		return nil
+	}
+
+	return s.flush(ctx)
+}
+
+// Close flushes anything that may be left in the underlying buffer and
+// terminates all background operations of the sender. Close is thread safe but
+// should only be called once no more calls to Send are needed; after Close has
+// been called any subsequent calls to Send will error while subsequent calls
+// to Close will no-op.
+func (s *sender) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	defer s.cancel()
+
+	if s.closed {
+		return nil
+	}
+	s.closed = true
+
+	if len(s.buffer) > 0 {
+		if err := s.flush(s.ctx); err != nil {
+			return errors.Wrap(err, "flushing buffer")
+		}
+	}
+
+	return nil
+}
+
+func (s *sender) timedFlush() {
+	s.mu.Lock()
+	s.timer = time.NewTimer(s.opts.FlushInterval)
+	s.mu.Unlock()
+	defer s.timer.Stop()
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-s.timer.C:
+			s.mu.Lock()
+			if len(s.buffer) > 0 && time.Since(s.lastFlush) >= s.opts.FlushInterval {
+				if err := s.flush(s.ctx); err != nil {
+					s.opts.Local.Send(message.NewErrorMessage(level.Error, err))
+				}
+			}
+			_ = s.timer.Reset(s.opts.FlushInterval)
+			s.mu.Unlock()
+		}
+	}
+}
+
+func (s *sender) flush(ctx context.Context) error {
+	if err := s.write(ctx, s.buffer); err != nil {
+		return errors.Wrap(err, "writing log")
+	}
+
+	s.buffer = []LogLine{}
```

Comment:
could it make sense to do something like
```go
s.buffer = s.buffer[:0]
```
to hold on to the allocated array?

Reply:
I think the only risk there is that we hold on to a large array that we may not end up using, wasting memory.